dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/common/DimensionSpecificationService.java

```diff
@@ -0,0 +1,195 @@
+                        dimensionIdWithOffset.getUid() ) )
+                    .orElseThrow( () -> new IllegalArgumentException(
+                        "Program stage " + programStageWithOffset.getUid() + " is not defined in program "
+                            + programWithOffset.getUid() ) );
+            }
+            else
+            {
+                throw new IllegalArgumentException( "Specified program " + programWithOffset + " does not exists" );
+            }
+        }
+        else if ( dsUids.hasProgram() )
+        { // example: {programUid}.PE
+
+            Program program = programOptional
+                .orElseThrow( () -> new IllegalArgumentException(
+                    ("Specified program " + programWithOffset + " does not exists") ) );
+
+            return DimensionSpecification.of(
+                DimensionSpecification.ElementWithOffset.of( program, programWithOffset.getOffset() ),
+                null,
+                dimensionIdWithOffset.getUid() );
+
+        }
+        else
+        {
+            return DimensionSpecification.of( null, null, dsUids.getDimension().getUid() );
+        }
+    }
+
+    private Optional<ProgramStage> extractProgramStageIfExists( Program program, String programStageUid )
+    {
+        return program.getProgramStages().stream()
+            .filter( programStage -> programStage.getUid().equals( programStageUid ) )
+            .findFirst();
+    }
+
+    @Data
+    @RequiredArgsConstructor( staticName = "of" )
+    private static class DimensionSpecificationUids
+    {
+
+        private final UidWithOffset program;
+
+        private final UidWithOffset programStage;
+
+        private final UidWithOffset dimension;
+
+        boolean hasProgram()
+        {
+            return Objects.nonNull( program );
+        }
+
+        boolean hasProgramStage()
+        {
+            return Objects.nonNull( programStage );
+        }
+
+        static DimensionSpecificationUids of( String fullDimensionId )
+        {
+            List<UidWithOffset> uidWithOffsets = parseFullDimensionId( fullDimensionId );
+
+            if ( uidWithOffsets.size() == 3 )
+            {
+                return DimensionSpecificationUids.of( uidWithOffsets.get( 0 ), uidWithOffsets.get( 1 ),
+                    uidWithOffsets.get( 2 ) );
+            }
+            if ( uidWithOffsets.size() == 2 )
+            {
+                return DimensionSpecificationUids.of( uidWithOffsets.get( 0 ), null, uidWithOffsets.get( 1 ) );
+
+            }
+            if ( uidWithOffsets.size() == 1 )
+            {
+                return DimensionSpecificationUids.of( null, null, uidWithOffsets.get( 0 ) );
+            }
+
+            throw new IllegalArgumentException( "Malformed parameter: " + fullDimensionId );
+
+        }
+
+        private static List<UidWithOffset> parseFullDimensionId( String fullDimensionId )
```

Comment:
I would add a Javadoc here explaining how many elements can result from the split of `fullDimensionId`, and examples of valid `fullDimensionId` that this method will accept.